rewrite-toml/src/main/java/org/openrewrite/toml/SemanticallyEqual.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2025 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.toml;
+
+import org.jspecify.annotations.Nullable;
+import org.openrewrite.toml.tree.Toml;
+import org.openrewrite.toml.tree.TomlKey;
+
+import java.util.List;
+
+/**
+ * Recursively check the equality of TOML elements.
+ * Ignores whitespace and formatting differences.
+ */
+public class SemanticallyEqual {
+
+    private SemanticallyEqual() {
+    }
+
+    public static boolean areEqual(Toml first, Toml second) {
+        SemanticallyEqualVisitor sme = new SemanticallyEqualVisitor();
+        sme.visit(first, second);
+        return sme.areEqual;
+    }
+
+    @SuppressWarnings("ConstantConditions")
+    private static class SemanticallyEqualVisitor extends TomlVisitor<Toml> {
+
+        boolean areEqual = true;
+
+        @Override
+        public Toml visitDocument(Toml.Document document, Toml other) {
+            if (document == other) {
+                return null;
+            }
+            if (!(other instanceof Toml.Document)) {
+                areEqual = false;
+                return null;
+            }
+            Toml.Document otherDocument = (Toml.Document) other;
+            visitList(document.getValues(), otherDocument.getValues());
```

Comment:
I'm suprised here by the explicit calls to `visitList` rather than the more established `super.visit` which would then also call `visitValues`. What was the thinking here? I'm inclined to think using established patterns of super visits would make it easier to reason about which methods get called when.